discojs/src/models/gpt/model.ts

```diff
@@ -194,44 +177,70 @@ function prepareIdx (idx: tf.TensorLike): tf.Tensor2D {
+      const idxNew = sequenceBatch.concat(idxNext, 1)
+      tf.dispose(sequenceBatch)
+      sequenceBatch = idxNew
       tf.dispose(idxNext)
     }
-    const idxArr = await idx.array()
-    tf.dispose(idx)
+    const idxArr = await sequenceBatch.array()
+    tf.dispose(sequenceBatch)
     return idxArr
   }
 
-  private generateOnce (model: tf.LayersModel, idx: tf.Tensor2D, config: GenerateConfig): tf.Tensor2D {
+  /**
+   * Generate the next token for each input sequence in the batch.
+   * In other words, takes an input tensor of shape (batch size B, prompt length T) and returns a tensor of shape (B, T+1)
+   * 
+   * @param sequenceBatch input token sequenceBatch of shape (B, T). T is truncated to the model's block size
+   * @param config generation config: maxNewTokens, temperature, doSample, topk
+   * @returns the next token predicted by the model for each input sequence in the batch
+   */
+  private generateOneToken (sequenceBatch: tf.Tensor2D, config: GenerationConfig): tf.Tensor2D {
     const idxNext = tf.tidy(() => {
-      // slice input tokens if longer than context length
+      // slice input sequenceBatch if longer than context length
       const blockSize = this.config.blockSize
-      idx = idx.shape[1] <= blockSize
-      ? idx : idx.slice([0, idx.shape[1] - blockSize])
-
-      const output = model.predict(idx)
+      // Truncate the prompt lengths to the max context length (batch size B, prompt length <= T)
+      sequenceBatch = sequenceBatch.shape[1] <= blockSize ? sequenceBatch : sequenceBatch.slice([0, sequenceBatch.shape[1] - blockSize])
+      const output = this.predict(sequenceBatch) // (B, T, vocab_size)
       if (Array.isArray(output)) throw new Error('The model outputs too multiple values')
       if (output.shape.length !== 3) throw new Error('The model outputs wrong shape')
-      const logits = output as tf.Tensor3D
+      const logits = output as tf.Tensor3D 
         
       const logitsScaled = logits
-        .slice([0, idx.shape[1] - 1, 0])
-        .reshape([logits.shape[0], logits.shape[2]])
-        .div<tf.Tensor2D>(tf.scalar(config.temperature))
-      const probs = logitsScaled.softmax(-1)
+        .slice([0, sequenceBatch.shape[1] - 1, 0]) // take the logits at the last position (B, 1, vocab_size)
+        .reshape([logits.shape[0], logits.shape[2]]) // (B, vocab_size)
+        .div<tf.Tensor2D>(tf.scalar(config.temperature)) // the higher the temperature, the more random the output
+      const probs = logitsScaled.softmax(-1) // get the token probabilities (B, vocab_size) 
       if (config.doSample) {
-        return tf.multinomial(probs, 1) as tf.Tensor2D
+        // returns topk biggest values among the `vocab_size` probabilities and the corresponding tokens indices 
+        // both shapes are (B, config.topk)
+        const { values: topkProbs, indices: topkTokens } = tf.topk(probs, config.topk);
+        // sample an index from the top-k probabilities
```

Comment:
top k first filters out every value except the first k and then use softmax to get a probability distribution (so all the probability mass distributed over the tokens dropped is being reallocated on the first top k). I include top-k sampling because it's a very common strategy in language generation